% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/RVFamSq.R
\name{RV_FamSq}
\alias{RV_FamSq}
\title{Rare Variant-Family-Based Score Test for Quantitative Traits}
\usage{
RV_FamSq(genofile, phenofile, maffile, parafile, covar_col, trait_col, out,
  kin, start_par = NULL)
}
\arguments{
\item{genofile}{referring to the path of the genotype file. The columns in
the file represent family IDs, individual IDs and genotypes of each
individual. The file should not have header.}

\item{phenofile}{referring to the path of the phenotype file. The first five
columns in the file should be family IDs,individual IDs, father IDs, mother
IDs, and sex, respectively. The rest columns of the file are covariates and
trait. The covariates may be multi-columns and trait should be one column.
The file should not have header.}

\item{maffile}{the path to the .sfs file describing the information of
variants in the genotype file.}

\item{parafile}{the path to the file saving parameters estimated in null
model. If the file is not existed, the package will estimate the parameters
using the available data and save the values of the parameters in parafile.}

\item{covar_col}{a vector defining the index of columns of covariates in
\code{phenofile}.}

\item{trait_col}{an integer number defining the column of interested trait
in \code{phenofile}.}

\item{out}{the directory that save the results. RVFamSq produces output file
with the name of the gene, score, p-value of the gene, number of sample size
and number of families}

\item{kin}{a kinship matrix calculated from the pedigree using the
“kinship2” package}

\item{start-par}{an optional argument that defines the starting values of
the parameters fitting. User can define specific values to start the
parameters fitting, otherwise, the starting values will generate randomly.}
}
\value{
\item{results}{a data frame results containing the name of the gene,
the value of the score and p-value write into the file named by the name of
the gene under the directory defined by the argument \code{out}.}
\item{paras}{a list of parameters estimated by maximizing the likelihood.
The values of the paras save to the file defined by the argument
\code{parafile}”. If the \code{parafile} is existed, the values of the paras
load from the file \code{parafile}.}
}
\description{
A regional association analysis of rare variants and quantitative traits in
family data
}
\details{
%% ~~ If necessary, more details than the description above ~~
}
\examples{

library (kinship2)
library (RVFamSq)

##  Define variables pointing to the path to the example data
data_dir<-system.file("data", package = "RVFamSq")
maffile<-paste0(data_dir,"/test.sfs")
genofile<-paste0(data_dir, "test.geno")
phenofile<-paste0(data_dir,"/test.pheno")

##  Define output directory that save the results
out<-"/directory-to-save-the-results"

## Define the files that save the parameters estimated under the null model.
## If the file is not existed, the package will estimate the parameters based on the available data.
parafile<-"/ directory-to-example-data /paras.rds"

## Load phenotype data and calculate the kinship matrix of the pedigree.
ped_pheno<-read.table(phenofile)
kin_pre<-data.frame(id=ped_pheno[,2], mom=ped_pheno[,4], dad=ped_pheno[,3], sex=ped_pheno[,5])
tped <- with(kin_pre, pedigree(id, dad, mom, sex, famid=ped_pheno[,1]))
kin <- kinship(tped)

## Run RVFamsSq package and calculate the statistical score of the interested gene.
RV_FamSq(pedfile, phenofile, maffile, parafile, c(5,7), 6, out,kin)

}
\references{
Chen, W.-M., and Abecasis, G.R. (2007). Family-Based Association
Tests for Genomewide Association Scans. Am. J. Hum. Genet. 81, 913–926.\cr
}
\author{
Zhihui Zhang and Suzanne Leal
}
